File: Pieces/Queen.mjs
import Square from "../Game/Square.mjs";
import Piece from "./Piece.mjs";

export default class Queen extends Piece {
  constructor(isWhite) {
    super(isWhite);
  }

  move = (board, initialSquare, destinationSquare) => {
    if (destinationSquare.piece?.isWhite === this.isWhite) return false;
    return this.isValidPath(board, initialSquare, destinationSquare);
  };

  pieceSymbol = () => {
    return this.isWhite ? "♛" : "♕";
  };

  isValidPath = (board, initialSquare, destinationSquare) => {
    let stepX = destinationSquare.x > initialSquare.x ? 1 : -1;
    let stepY = destinationSquare.y > initialSquare.y ? 1 : -1;

    let x = initialSquare.x + stepX;
    let y = initialSquare.y + stepY;

    if (this.isHorizontalMove(initialSquare, destinationSquare)) {
      let row = initialSquare.x;

      while (y != destinationSquare.y) {
        if (board[row][y].piece) return false; // There is a piece blocking the path
        y += stepY;
      }

      return true;
    }

    if (this.isStraightMove(initialSquare, destinationSquare)) {
      let col = initialSquare.y;

      while (x != destinationSquare.x) {
        if (board[x][col].piece) return false; // There is a piece blocking the path
        x += stepX;
      }

      return true;
    }

    if (this.isDiagonalMove(initialSquare, destinationSquare)) {
      while (x !== destinationSquare.x && y !== destinationSquare.y) {
        if (board[x][y].piece) {
          return false; // There is a piece blocking the path
        }
        x += stepX;
        y += stepY;
      }

      return true;
    }

    return false;
  };

  isDiagonalMove = (initialSquare, destinationSquare) => {
    let dx = Math.abs(initialSquare.x - destinationSquare.x);
    let dy = Math.abs(initialSquare.y - destinationSquare.y);

    return dx === dy;
  };

  isHorizontalMove = (initialSquare, destinationSquare) => {
    return initialSquare.x === destinationSquare.x;
  };

  isStraightMove = (initialSquare, destinationSquare) => {
    return initialSquare.y === destinationSquare.y;
  };
}

File: Game/Pieces/Queen.py
from .Piece import Piece
from ..GUI import Utils

class Queen(Piece):
    def __init__(self, is_white):
        super().__init__(is_white)

    def piece_image(self):
        if self.is_white:
            return Utils.pieces["wq"]
        else:
            return Utils.pieces["bq"]
        
    def piece_notation():
        return "Q"

    def move(self, board, initial_square, destination_square):       
        return not self.has_friendly_piece(destination_square) and self.is_valid_path(board, initial_square, destination_square)

    def is_valid_path(self, board, initial_square, destination_square):
        step_x = 1 if destination_square.x > initial_square.x else -1
        step_y = 1 if destination_square.y > initial_square.y else -1

        x = initial_square.x + step_x
        y = initial_square.y + step_y

        if self.is_horizontal_move(initial_square, destination_square):
            row = initial_square.x
            while y != destination_square.y:
                if board[row][y].piece:
                    return False  # There is a piece blocking the path
                y += step_y
            return True

        if self.is_straight_move(initial_square, destination_square):
            col = initial_square.y
            while x != destination_square.x:
                if board[x][col].piece:
                    return False  # There is a piece blocking the path
                x += step_x
            return True

        if self.is_diagonal_move(initial_square, destination_square):
            while x != destination_square.x and y != destination_square.y:
                if board[x][y].piece:
                    return False  # There is a piece blocking the path
                x += step_x
                y += step_y
            return True

        return False
